### ISSUE TITLE
Move all the field-type modules outside of the field module directory

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#557]
Follow up to Follow up to https://github.com/backdrop/backdrop-issues/issues/486

All the field-type modules like 'number' and 'text' currently live in `core/modules/field/modules` which seems a little odd.  Let's move these to the root of the modules directory like all the others.

**Comment by BWPanda** [comment]
Made a new [PR](https://github.com/backdrop/backdrop/pull/2863) based on @hellomrcat's one. There were still some files from the Options module in `fields/modules`. Not sure if they were an oversight or have been added since the PR was made. In any case, PR updated to apply to latest code.

**Comment by bugfolder** [comment]
Tested, WFM (i.e., fields still work). Code reviewed, LGTM.

**Comment by argiepiano** [comment]
When I manually apply this patch in my local, the files and folders are not moved. Something is wrong with this patch, but I don't know enough about patch files to find out what the problem is.

**Comment by bugfolder** [comment]
Ditto. They're in the right place in the original fork, but when I download the code from the PR, they're not.

**Comment by argiepiano** [comment]
The only actual noticeable change after applying patch is the code change in line 9 of `options.test`.

**Comment by BWPanda** [comment]
That must just be something about the way patches/diff's work, 'cause everything looks fine in the vrach I'm merging: https://github.com/BWPanda/backdrop/tree/issue-557/core/modules

**Comment by bugfolder** [comment]
Yes, I saw they were correct in the fork. So how about we go ahead and mark it RTBC, but then ask the PtB to verify that the changes "took" when the PR is merged? (Since files have been moved around before in PRs, I'm guessing they will "take".)

**Comment by indigoxela** [comment]
Hm... I don't see any code that actually moves directories, but maybe I just missed that.

To my understanding, the PR just move the directories, with one single change in one require_once. So updating core will lead to duplicate module directories, as core updates just unpack the new core over the old one. (See also  #4285 re core updates and obsolete files.)

Can anyone point me to the code that takes care of the directories in the old location?

**Comment by bugfolder** [comment]
> So updating core will lead to duplicate module directories, as core updates just unpack the new core over the old one

Wow, did not know that.

So maybe a `hook_update_N()` that looks for and removes the old directories? Would that be the (or a) way of handling this?

**Comment by BWPanda** [comment]
Does this help? https://github.com/backdrop/backdrop/pull/537/files That was the merged PR for https://github.com/backdrop/backdrop-issues/issues/204

**Comment by indigoxela** [comment]
> Does this help? ...

I'm not sure. That also happened before core updates via UI got implemented. And it's only about functional tests. Back in 2015, when this issue was opened, moving modules around may have been an easy task, but since then a lot of things happened.

What exactly happens, if one updates via UI and the modules are in two places - even if it's only until an update hook runs to remove them? - And we also have to take care of the system table, where the module paths have to get updated.
I'm pretty sure, Backdrop nags, but didn't test that yet.

**Comment by BWPanda** [comment]
I think I found the problem: you either need to move files using your OS, then **add the new location** to git **and remove the old**, or use git itself to move files (which handles all that for you): https://githowto.com/moving_files

I'll update the PR soon. I don't think update hooks, etc. are necessary.

**Comment by indigoxela** [comment]
> I don't think update hooks, etc. are necessary.

I suspect, you missed the base problem, so again: when updating via UI, the new code gets unpacked over the old code, so modules will be in both locations after that.

And the system table contains the old paths in system.filename - like `core/modules/field/modules/field_sql_storage/field_sql_storage.module`.

So there **has** to be code to fix both. The problem has nothing to do with git, it's about updating via UI.

**Comment by BWPanda** [comment]
Ah, very well.

**Comment by bugfolder** [comment]
> What exactly happens, if one updates via UI and the modules are in two places - even if it's only until an update hook runs to remove them?

Badness and confusion can happen (especially if the modules aren't truly identical), and https://github.com/backdrop/backdrop-issues/issues/5464 now puts a warning on the status page.
